### ISSUE TITLE
InputInteractionContext setTimeout seems to be working weird ?

### ISSUE THREAD

**Issue by gotenxds** [Unity-Technologies/InputSystem#714]
Hello, I'm experimenting with creating custom interactions and I'm having issues with setTimeout
This may be by lack of documentation about this but I think it is acting weird

take this code for example: 

`

        public void Process(ref InputInteractionContext context)
        {
            Debug.Log("Proccessing");

            context.SetTimeout(1);
        }
`

What I would expect to happen is that every 1 second I would get the output `Proccessing` in the console
What I get instead is just 2, the intial invocation of the action and the timeout

![image](https://user-images.githubusercontent.com/3519520/60760556-dd700280-a03f-11e9-8930-d3a9ee3430c0.png)

So I tought, maybe setting the timeout every time causing some issue and actually resets the timer ?

`

        private bool timeoutInitialized;

        public void Process(ref InputInteractionContext context)
        {
            Debug.Log("Proccessing");

            if (timeoutInitialized)
            {
                context.SetTimeout(1);
                timeoutInitialized = true;
            }
        }
`

But again, I get the same result, just two calls.

**Comment by sttz** [comment]
I think I hit the same bug, it's currently not possible to set a new timeout when another timeout has just been triggered.

The issue is in `InputManager.ProcessStateChangeMonitorTimeouts()`. When a timeout expires, it calls `IInputStateChangeMonitor.NotifyTimerExpired()` inline, which in turn immediately calls `IInputInteraction.Process()`. If you try to set a new timeout during that `Process` call, it will get added to `m_StateChangeMonitorTimeouts` while `ProcessStateChangeMonitorTimeouts()` is still iterating over them.

The issue is that `ProcessStateChangeMonitorTimeouts()` caches the length of the array, meaning it will ignore newly added timeouts and its compaction strategy then truncates the array, removing them.

To test this, I removed the length caching and the timeouts started to work as expected:

```cs
        private void ProcessStateChangeMonitorTimeouts()
        {
            if (m_StateChangeMonitorTimeouts.length == 0)
                return;

            // Go through the list and both trigger expired timers and remove any irrelevant
            // ones by compacting the array.
            // NOTE: We do not actually release any memory we may have allocated.
            var currentTime = m_Runtime.currentTime - InputRuntime.s_CurrentTimeOffsetToRealtimeSinceStartup;
            var remainingTimeoutCount = 0;
            for (var i = 0; i < m_StateChangeMonitorTimeouts.length; ++i)
            {
                // If we have reset this entry in RemoveStateChangeMonitorTimeouts(),
                // skip over it and let compaction get rid of it.
                if (m_StateChangeMonitorTimeouts[i].control == null)
                    continue;

                var timerExpirationTime = m_StateChangeMonitorTimeouts[i].time;
                if (timerExpirationTime <= currentTime)
                {
                    var timeout = m_StateChangeMonitorTimeouts[i];
                    timeout.monitor.NotifyTimerExpired(timeout.control,
                        currentTime, timeout.monitorIndex, timeout.timerIndex);

                    // Compaction will get rid of the entry.
                }
                else
                {
                    // Rather than repeatedly calling RemoveAt() and thus potentially
                    // moving the same data over and over again, we compact the array
                    // on the fly and move entries in the array down as needed.
                    if (i != remainingTimeoutCount)
                        m_StateChangeMonitorTimeouts[remainingTimeoutCount] = m_StateChangeMonitorTimeouts[i];
                    ++remainingTimeoutCount;
                }
            }

            m_StateChangeMonitorTimeouts.SetLength(remainingTimeoutCount);
        }
```

**Comment by Rene-Damm** [comment]
Good catch. Thank you for the fix @sttz. I'll add a PR with the fix.

**Comment by sttz** [comment]
@Rene-Damm Great, thanks!

Note that the same issue exists with `RemoveStateChangeMonitorTimeout()` (e.g. called through a `StopTimeout()` in the `IInputInteraction.Process()` callback), in which case an arbitrary timeout could be removed inadvertently. This fix doesn't apply there and fixing it isn't as simple. 😟

**Comment by Rene-Damm** [comment]
>Note that the same issue exists with RemoveStateChangeMonitorTimeout() (e.g. called through a StopTimeout() in the IInputInteraction.Process() callback), in which case an arbitrary timeout could be removed inadvertently. This fix doesn't apply there and fixing it isn't as simple. 😟

I'll take a look. Thanks for the heads-up :)

**Comment by Rene-Damm** [comment]
Should be fixed by #816.

>Note that the same issue exists with RemoveStateChangeMonitorTimeout() (e.g. called through a StopTimeout() in the IInputInteraction.Process() callback), in which case an arbitrary timeout could be removed inadvertently. This fix doesn't apply there and fixing it isn't as simple. 😟

Seems like the removal logic was missing a change that got lost along the way. It did `RemoveAt` whereas the processing logic that has the compaction code stated it doesn't do that. Also noticed that as monitors were removed, potential timeouts on the monitor were left in place.

Should be fixed in the PR.
